### PR TITLE
WIP (do not merge yet) pack for NAPALM library (multivendor network automation)

### PR DESCRIPTION
**Pull request by Mierdin**
## NAPALM

[NAPALM](https://github.com/napalm-automation) is a Python library to simplify and abstract some of the programmatic communication with network devices, making multi-vendor network automation a little easier. This pack introduces new capabilities that allow a StackStorm user to leverage NAPALM within StackStorm in the form of sensors, actions, and more.

### Status 

This is a work-in-progress, please do not merge for the time being.

### Description

This PR will introduce a new pack that leverages the NAPALM library to allow ST2 to perform multivendor network automation

### Checklist (tick everything that applies)

- [ ] Metadata: pack.yaml, icon, structure (required for new packs)
- [ ] Version bump (required for changed packs)
- [ ] Code linting (required, can be done after the PR checks)
- [ ] [Tests](https://docs.stackstorm.com/development/pack_testing.html) (not required but really recommended)

### CODE REVIEW

File: packs/napalm/config.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+---
```
Comment (LindsayHill) [inline packs/napalm/config.yaml]: Could you change to use `config.schema.yaml` rather than `config.yaml`

Comment (Mierdin) [inline packs/napalm/config.yaml]: What's the distinction? I am basing this off of the other packs I saw in st2contrib, which all use `config.yaml`

Comment (LindsayHill) [inline packs/napalm/config.yaml]: `config.schema.yaml` was introduced in v1.5 (https://stackstorm.com/2016/06/30/improved-pack-configuration-user-scoped-datastore-items-secure-secrets-store/). 

Both versions were supported for the last ~6 months or so. But the new pack management `st2 pack config` command will only work with `config.schema.yaml` (https://docs.stackstorm.com/latest/packs.html#configuring-a-pack). It is still technically possible to use older style configuration with new packs, but it is less user-friendly, and should be discouraged.

There should be very few remaining packs using `config.yaml`. I've made many changes over the last few weeks to migrate packs. As of yesterday, around 12 of 105 packs were using `config.yaml`. Probably fewer by now.

Comment (Mierdin) [inline packs/napalm/config.yaml]: Ah, okay. Fixed in 74e3429

**Comment by LindsayHill** [comment]
For migrations from `config.yaml` to `config.schema.yaml`, you'll want to rename `config.yaml` to `napalm.yaml.example`, and create a new `config.schema.yaml` file. The format looks quite different for `config.schema.yaml`.

Flat configs are pretty easy to create `config.schema.yaml` for, but it's a bit more complicated for a setup like yours where there is an array. Part of it will look like this:

```
  driver:
    description: "NAPALM driver type - e.g. junos, eos."
    type: "string"
    required: true
 host:
    description: "IP or hostname of device"
    type: "string"
    required: true
username:
    description: "Username to access device"
    type: "string"
    required: true
  password:
    description: "Password for device"
    type: "string"
    required: true
    secret: true
 port:
    description: "SSH port - default 22"
    type: "integer"
    default: 22
```

That will need to wrapped in a higher-level array object. See something like the cubesensors config.schema for an example.

If you have a "<pack_name>.yaml.example" file, then CI tests will check that the example config matches with the config schema.

Oh and I guess we'll need to move this across to ST2-Exchange shortly.

**Comment by emedvedev** [comment]
> Oh and I guess we'll need to move this across to ST2-Exchange shortly.
Preferably. :) No rush though. You can use the [incubator repo](https://github.com/stackstorm-exchange/exchange-incubator), or just bootstrap a new one with CI and everything.

**Comment by Mierdin** [comment]
@emedvedev @LindsayHill yeah the move to exchange is in the back of my mind - when this is more or less looking like a finished pack, I'll make sure that's opened - this is just temporary

**Comment by Mierdin** [comment]
@LindsayHill How's this?

`napalm.yaml.example`:
```
---
devices:
- driver: junos
  hostname: 192.168.50.51
  username: root
  password: Juniper
  port: 22
- driver: junos
  hostname: 192.168.50.52
  username: root
  password: Juniper
  port: 22
```

`config.schema.yaml`:
```
---
devices:
  description: "The array of devices"
  required: true
  type: "array"
  items:
    type: "object"
    required: true
    properties:
      driver:
        description: "Driver for the device"
        type: "string"
        required: true
      hostname:
        description: "Hostname for the device"
        type: "string"
        required: true
      username:
        description: "Username for the device"
        type: "string"
        required: true
      password:
        description: "Password for the device"
        type: "string"
        required: true
        secret: true
      port:
        description: "Port for the device"
        type: "integer"
        required: true
```

**Comment by LindsayHill** [comment]
I _think_ that looks OK. I think there's a `make` target to test configs.

**Comment by Mierdin** [comment]
Cool, I was able to run `FORCE_CHECK_PACK=napalm make configs-check` and test a valid configuration. https://github.com/StackStorm/st2contrib/pull/634/commits/0fd2f1e1b45d2c372d8312bc47945e6b56b3f576 should be a better commit wrt this

**Comment by Mierdin** [comment]
I am closing this PR, as the new home for this pack is present in it's own repo under the exchange.

Please see https://github.com/StackStorm-Exchange/stackstorm-napalm/pull/2 for updates on the initial contribution of this pack.

**Pull request closed without merging**